File: Day4/Day4.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

class Day4 {

    public static int check(String[] pairs) {
        // Just check the basic logic
        int x = (int) Integer.valueOf(pairs[0].split("-")[0]);
        int y = (int) Integer.valueOf(pairs[0].split("-")[1]);
        int a = (int) Integer.valueOf(pairs[1].split("-")[0]);
        int b = (int) Integer.valueOf(pairs[1].split("-")[1]);

        if ((x <= a && y >= b) || (x >= a && y <= b)) {
            return 1;
        }
        return 0;
    }


    public static int overlap(String[] pairs) {
        // Alter the logic a bit
        int x = (int) Integer.valueOf(pairs[0].split("-")[0]);
        int y = (int) Integer.valueOf(pairs[0].split("-")[1]);
        int a = (int) Integer.valueOf(pairs[1].split("-")[0]);
        int b = (int) Integer.valueOf(pairs[1].split("-")[1]);

        if (Integer.max(x, y) < Integer.min(a, b) || Integer.min(x, y) > Integer.max(a, b)) {
            return 0;
        }
        return 1;
    }
    
    public static int problem1(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            int total = 0;
            for (String line: lines) {
                total += check(line.split(",")); // Return 0 if false, 1 if true, probably should've done this in python too smh
            }
            
            return total;
        }
        finally
        {
            s.close();
        }
    }

    public static int problem2(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            int total = 0;
            for (String line: lines) {
                total += overlap(line.split(","));
            }
            
            return total;
        }
        finally
        {
            s.close();
        }
    }
    
    public static void main(String[] args) {
        try {
            System.out.println(problem1("data.txt"));
            System.out.println(problem2("data.txt"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Day4/Day4.py
def problem1(file: str) -> str:
    """
    Solution to problem1
    """
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    # I think the best way to go about this will be to check if the first number is smaller then the other and so on
    def check(pairs: list) -> bool:
        """
        Check pairs
        """
        x, y = pairs[0].split("-")
        a, b = pairs[1].split("-")
        return (int(x) <= int(a) and int(y) >= int(b)) or (
            int(x) >= int(a) and int(y) <= int(b)
        )

    total = 0
    for line in lines:
        if check(line.split(",")):
            total += 1

    return total


def problem2(file: str) -> str:
    """
    Solution to problem2
    """
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    def overlap(pairs: list) -> bool:
        """
        Check if the pairs overlap, we just have to edit my code before a bit
        """
        x, y = pairs[0].split("-")
        a, b = pairs[1].split("-")
        x = int(x)
        y = int(y)
        a = int(a)
        b = int(b)

        # For this logic, we can just check if the bigger number of x and y are smaller than the biggest of a or b and reverse it
        if max(x, y) < min(a, b) or min(x, y) > max(a, b):
            return False
        return True

    total = 0
    for line in lines:
        if overlap(line.split(",")):
            total += 1

    return total


print(problem1("data.txt"))
print(problem2("data.txt"))
